pricewars_merchant.py

```diff
@@ -68,13 +73,32 @@ def calculate_id(token: str) -> str:
         return base64.b64encode(hashlib.sha256(token.encode('utf-8')).digest()).decode('utf-8')
 
     def run(self):
+        # initial random sleep to avoid starting merchants in sync
+        time.sleep(2 * random.random())
+
         start_time = time.time()
+        cnt = 1
         while True:
+            il = self.settings['update interval']
+            lb = self.settings['interval_lower_bound_relative']
+            ub = self.settings['interval_upper_bound_relative']
+
             if self.state == 'running':
                 self.update_offers()
-            # Waiting for the length of the update interval minus the execution time
-            time.sleep(self.settings['update interval'] -
-                       ((time.time() - start_time) % self.settings['update interval']))
+
+            # determine required sleep length for next interval
+            rdm_interval_length = (il * lb) + (random.random() * il * (ub - lb))
+            # calculate next expected update timespamp (might be in the past)
+            next_update_ts = start_time + il * (cnt-1) + rdm_interval_length
+            sleep_time = next_update_ts - time.time()
+
+            if sleep_time <= 0:
+                # short random sleep to catch up with the intervals,
+                # but try not to DDos the marketplace
+                sleep_time = (il * 0.05) + (random.random() * il * 0.1)
```

Comment:
Don't really understand this calculation. Would it be enough, if you just use `sleep_time = max(next_update_ts - time.time(), some_constant)`?

Reply:
I think the `if` is more verbose and even the waiting times in case of "catching up" should be somewhat random to avoid several merchant DDos'ing the marketplace.

But you're absolutely right. There are nice ways to write that.I'll take a look at it again.

Reply:
I like the verbosity, since your first comment appeared to overlook the problem, that some blocking component might lead to a merchant not fully using the allowed rates.

As such, the additional if clause is more explicit.

So far done with your review.